blockly/blocks/component_variable.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * @fileoverview Variable input field of a specific component type.
+ * @author fraser@google.com (Neil Fraser)
+ */
+'use strict';
+
+goog.provide('Blockly.Blocks.ComponentFieldVariable');
+
+goog.require('Blockly.Blocks');
+//goog.require('Blockly.FieldVariable');
+
+
+/**
+ * Class for a variable's dropdown field.
+ * @param {?string} varname The default name for the variable.  If null,
+ *     a unique variable name will be generated.
+ * @param {Function=} opt_validator A function that is executed when a new
+ *     option is selected.  Its sole argument is the new option value.
+ * @extends {Blockly.FieldVariable}
+ * @constructor
+ */
+Blockly.Blocks.ComponentFieldVariable = function(varname, component_type, opt_validator) {
+  this.component_type = component_type
+  //override the dropdownCreate to use for this field
+  this.dropdownCreate = Blockly.Blocks.ComponentFieldVariable.dropdownCreateComponents;
+  Blockly.Blocks.ComponentFieldVariable.superClass_.constructor.call(this,
+      varname, opt_validator);
+  this.setValue(varname || '');
+};
+goog.inherits(Blockly.Blocks.ComponentFieldVariable, Blockly.FieldVariable);
+
+/**
+ * Install this dropdown on a block.
+ * @param {!Blockly.Block} block The block containing this text.
+ */
+Blockly.Blocks.ComponentFieldVariable.prototype.init = function(block) {
+  if (this.sourceBlock_) {
+    // Dropdown has already been initialized once.
+    return;
+  }
+  Blockly.Blocks.ComponentFieldVariable.superClass_.init.call(this, block);
+};
+
+Blockly.Blocks.ComponentFieldVariable.ComponentVariables = function(root, component_type) {
+  var blocks;
+  if (root.getDescendants) {
+    // Root is Block.
+    blocks = root.getDescendants();
+  } else if (root.getAllBlocks) {
+    // Root is Workspace.
+    blocks = root.getAllBlocks();
+  } else {
+    throw 'Not Block or Workspace: ' + root;
+  }
+  var variableHash = Object.create(null);
+  // Iterate through every block and add each variable to the hash.
+  for (var x = 0; x < blocks.length; x++) {
+    if (blocks[x].getComponentName && blocks[x].getVars && 
+        (blocks[x].getComponentName() ==  component_type)) {
+      var blockVariables = blocks[x].getVars();
+      for (var y = 0; y < blockVariables.length; y++) {
+        var varName = blockVariables[y];
+        // Variable name may be null if the block is only half-built.
+        if (varName) {
+          variableHash[varName.toLowerCase()] = varName;
+        }
+      }
+    }
+  }
+  // Flatten the hash into a list.
+  var variableList = [];
+  for (var name in variableHash) {
+    variableList.push(variableHash[name]);
+  }
+  return variableList;
+};
+
+
+/**
+ * Return a sorted list of variable names for variable dropdown menus.
+ * Include a special option at the end for creating a new variable name.
+ * @return {!Array.<string>} Array of variable names.
+ * @this {!Blockly.FieldVariable}
+ */
+Blockly.Blocks.ComponentFieldVariable.dropdownCreateComponents = function() {
```

Comment:
Could we just overwrite the superclass method `dropdownCreate`? This would also remove the need to have that `this.dropdownCreate = Blockly.Blocks.ComponentFieldVariable.dropdownCreateComponents;` in the constructor.